rust-runtime/aws-smithy-mocks-experimental/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Automatically managed default lints */
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+/* End of automatically managed default lints */
+use std::fmt::{Debug, Formatter};
+use std::future::Future;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use aws_smithy_runtime_api::box_error::BoxError;
+use aws_smithy_runtime_api::client::interceptors::context::{
+    BeforeDeserializationInterceptorContextMut, BeforeSerializationInterceptorContextMut, Error,
+    FinalizerInterceptorContextMut, Input, Output,
+};
+use aws_smithy_runtime_api::client::interceptors::Intercept;
+use aws_smithy_runtime_api::client::orchestrator::{HttpResponse, OrchestratorError};
+use aws_smithy_runtime_api::client::result::SdkError;
+use aws_smithy_runtime_api::client::runtime_components::RuntimeComponents;
+use aws_smithy_runtime_api::http::{Response, StatusCode};
+use aws_smithy_types::body::SdkBody;
+use aws_smithy_types::config_bag::{ConfigBag, Storable, StoreReplace};
+
+#[macro_export]
+macro_rules! mock {
+    ($operation: expr) => {
+        #[allow(unreachable_code)]
+        {
+            $crate::RuleBuilder::new(
+                || $operation(todo!()).as_input().clone().build().unwrap(),
+                || $operation(todo!()).send(),
+            )
```

Comment:
Add comments that the closures are purely for inferring the correct generic types in RuleBuilder.